File: Assets/Bundles/Shared/Resources/TS/Player/Player.lua
-- Compiled with unity-ts v2.1.0-75
local Flamework = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Flamework
local Network = require("Shared/TS/Network").Network
local Bin = require("Shared/TS/Util/Bin").Bin
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local Signal = require("Shared/TS/Util/Signal").Signal
local Player
do
	Player = setmetatable({}, {
		__tostring = function()
			return "Player"
		end,
	})
	Player.__index = Player
	function Player.new(...)
		local self = setmetatable({}, Player)
		return self:constructor(...) or self
	end
	function Player:constructor(nob, clientId, userId, username, usernameTag)
		self.nob = nob
		self.clientId = clientId
		self.userId = userId
		self.username = username
		self.usernameTag = usernameTag
		self.CharacterChanged = Signal.new()
		self.OnLeave = Signal.new()
		self.OnChangeTeam = Signal.new()
		self.bin = Bin.new()
		self.connected = true
	end
	function Player:SetTeam(team)
		local oldTeam = self.team
		self.team = team
		self.OnChangeTeam:Fire(team, oldTeam)
	end
	function Player:GetTeam()
		return self.team
	end
	function Player:SendMessage(message)
		if RunUtil:IsServer() then
			Network.ServerToClient.ChatMessage.Server:FireClient(self.clientId, message)
		else
			(Flamework.resolveDependency("Bundles/Client/Controllers/Global/Chat/ChatController@ChatController")):AddChatMessage(message)
		end
	end
	function Player:IsFriend()
		return false
	end
	function Player:IsBot()
		return false
	end
	function Player:Encode()
		local _object = {
			nobId = self.nob.ObjectId,
			clientId = self.clientId,
			userId = self.userId,
			username = self.username,
			usernameTag = self.usernameTag,
		}
		local _left = "teamId"
		local _result = self.team
		if _result ~= nil then
			_result = _result.id
		end
		_object[_left] = _result
		return _object
	end
	function Player:SetCharacter(entity)
		local _result = entity
		if _result ~= nil then
			_result = _result:GetDisplayName()
		end
		local _condition = _result
		if _condition == nil then
			_condition = "undefined"
		end
		print("setCharacter " .. _condition)
		self.Character = entity
		self.CharacterChanged:Fire(entity)
	end
	function Player:ObserveCharacter(observer)
		local bin = Bin.new()
		local cleanup = observer(self.Character)
		bin:Add(self.CharacterChanged:Connect(function(newPawn)
			local _result = cleanup
			if _result ~= nil then
				_result()
			end
			cleanup = observer(newPawn)
		end))
		self.bin:Add(bin)
		return bin
	end
	function Player:IsConnected()
		return self.connected
	end
	function Player:Destroy()
		self.connected = false
		self.bin:Clean()
		self.OnLeave:Fire()
		self.OnLeave:DisconnectAll()
	end
end
return {
	Player = Player,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Shared/Player/Player.ts
import { Dependency } from "@easy-games/flamework-core";
import { ChatController } from "Client/Controllers/Chat/ChatController";
import { PlayerController } from "Client/Controllers/Player/PlayerController";
import { PlayerService } from "Server/Services/Player/PlayerService";
import { CharacterEntity } from "Shared/Entity/Character/CharacterEntity";
import { Entity } from "Shared/Entity/Entity";
import { CoreNetwork } from "Shared/Network";
import { Team } from "../Team/Team";
import { Bin } from "../Util/Bin";
import { RunUtil } from "../Util/RunUtil";
import { Signal } from "../Util/Signal";

export interface PlayerDto {
	nobId: number;
	clientId: number;
	userId: string;
	username: string;
	usernameTag: string;
	teamId: string | undefined;
}

export class Player {
	/**
	 * The player controls this entity.
	 */
	public Character: CharacterEntity | undefined;
	/** Fired when the player's character changes. */
	public readonly CharacterChanged = new Signal<CharacterEntity | undefined>();
	/**
	 * Fired when the player disconnects from the server.
	 * Connections will automatically be disconnected when the player leaves.
	 */
	public readonly OnLeave = new Signal<void>();

	private team: Team | undefined;
	public readonly OnChangeTeam = new Signal<[team: Team | undefined, oldTeam: Team | undefined]>();

	private bin = new Bin();
	private connected = true;

	constructor(
		/**
		 * The GameObject representing the player.
		 */
		public readonly nob: NetworkObject,

		/**
		 * Unique network ID for the player in the given server. This ID
		 * is typically given to network requests as a way to identify the
		 * player to/from the server.
		 *
		 * This is not a unique identifier for the player outside of the
		 * server. For a completely unique ID, use `BWPlayer.clientId`
		 * instead.
		 */
		public readonly clientId: number,

		/**
		 * The player's unique ID. This is unique and unchanging per player.
		 *
		 * This should _not_ be used in network requests to identify the
		 * player. Use `clientId` for network requests.
		 */
		public readonly userId: string,

		/**
		 * The player's username. Non-unique, unless combined with `usernameTag`.
		 */
		public readonly username: string,

		/**
		 * The player's username tag. Append this value onto `username` for a
		 * unique username.
		 * ```ts
		 * const uniqueName = `${player.username}#${player.usernameTag}`;
		 * ```
		 */
		public readonly usernameTag: string,
	) {}

	public SetTeam(team: Team): void {
		const oldTeam = this.team;
		this.team = team;
		this.OnChangeTeam.Fire(team, oldTeam);
	}

	public GetTeam(): Team | undefined {
		return this.team;
	}

	public SendMessage(message: string): void {
		if (RunUtil.IsServer()) {
			CoreNetwork.ServerToClient.ChatMessage.Server.FireClient(this.clientId, message);
		} else {
			Dependency<ChatController>().AddChatMessage(message);
		}
	}

	/** Is player friends with the local player? */
	public IsFriend(): boolean {
		return false;
	}

	public IsBot(): boolean {
		return this.clientId < 0;
	}

	public Encode(): PlayerDto {
		return {
			nobId: this.nob.ObjectId,
			clientId: this.clientId,
			userId: this.userId,
			username: this.username,
			usernameTag: this.usernameTag,
			teamId: this.team?.id,
		};
	}

	public SetCharacter(entity: CharacterEntity | undefined): void {
		print("setCharacter " + tostring(entity?.id));
		this.Character = entity;
		this.CharacterChanged.Fire(entity);
	}

	public ObserveCharacter(observer: (entity: Entity | undefined) => CleanupFunc): Bin {
		const bin = new Bin();
		let cleanup = observer(this.Character);

		bin.Add(
			this.CharacterChanged.Connect((newCharacter) => {
				cleanup?.();
				cleanup = observer(newCharacter);
			}),
		);

		this.bin.Add(bin);
		return bin;
	}

	/**
	 * Is the player connected to the server?
	 */
	public IsConnected(): boolean {
		return this.connected;
	}

	public Destroy(): void {
		this.connected = false;
		this.bin.Clean();
		this.OnLeave.Fire();
		this.OnLeave.DisconnectAll();
	}

	public static FindByClientId(clientId: number): Player | undefined {
		if (RunUtil.IsServer()) {
			return Dependency<PlayerService>().GetPlayerFromClientId(clientId);
		} else {
			return Dependency<PlayerController>().GetPlayerFromClientId(clientId);
		}
	}
}
